### PR TITLE
Fix potential memory leak in ASN1_TIME_to_generalizedtime()

### PR DESCRIPTION
**Pull request by tmshort**
If ret is allocated, it may be leaked on error.

##### Checklist

- [ ] tests are added or updated

### CODE REVIEW

File: crypto/asn1/a_time.c [hunk 1]
```diff
@@ -96,11 +94,20 @@ ASN1_GENERALIZEDTIME *ASN1_TIME_to_generalizedtime(ASN1_TIME *t,
...
+       *out = ret;
+   return ret;
+
+ err:
+    if (out == NULL || *out != ret)
+        ASN1_GENERALIZEDTIME_free(ret);
+    return NULL;
 }
 
+
```
Comment (bernd-edlinger) [inline crypto/asn1/a_time.c]: extra line.

**Comment by FdaSilvaYY** [comment]
I don't like the way of this fix ...
Just adding a `alloc `pointer, to manage the internal allocation is more easy and simpler.

**Comment by tmshort** [comment]
I disagree, adding another variable, adds another variable, and doesn't really reduce complexity. Another change to note is that previously, on error, the *out pointer could have been modified to point to now freed memory. The fixed version only updates the *out pointer on success.

File: crypto/asn1/a_time.c (hunk 1 above)
Comment (tmshort) [inline crypto/asn1/a_time.c]: oops.

**Comment by tmshort** [comment]
ping?

**Comment by FdaSilvaYY** [comment]
I was thinking to a fix like this : https://github.com/openssl/openssl/commit/b8f1c116a357285ccb4905cd88c83f5076bafb52
Sono need to duplicate the test logic between alloc and free places.

**Comment by t-j-h** [comment]
+1

**Comment by richsalz** [comment]
OpenSSL_1_0_2-stable cdfb780
OpenSSL_1_1_0-stable ca0a7a9
master               4483e23

thanks!

**Comment by davidben** [comment]
FYI, the OpenSSL_1_0_2-stable cherry-pick appears to have a mistake:

```
        if (!(ret = ASN1_GENERALIZEDTIME_new()))
            goto err;
        if (out)
            *out = ret;
```

The `if (out)` block should not be there. This is handled by the `done` block. Otherwise, the function will continue to modify `*out` on error. The master and OpenSSL_1_1_0-stable versions do not have this.

(It's also missing the `const` in the final `BUF_strlcat` call, though that doesn't really matter.)

**Comment by bernd-edlinger** [comment]
Confirmed, thanks for pointing that out.
@davidben will you open a PR for this?

**Comment by mattcaswell** [comment]
I created a fix in #3002

**Pull request closed without merging**